crypto/fipsmodule/pqdsa/pqdsa.c

```diff
@@ -101,18 +104,31 @@ int PQDSA_KEY_set_raw_keypair_from_seed(PQDSA_KEY *key, CBS *in) {
     return 0;
   }
 
+  uint8_t *seed = OPENSSL_malloc(key->pqdsa->keygen_seed_len);
+  if (seed == NULL) {
+    OPENSSL_free(private_key);
+    OPENSSL_free(public_key);
+    return 0;
+  }
+
   // attempt to generate the key from the provided seed
   if (!key->pqdsa->method->pqdsa_keygen_internal(public_key,
                                                  private_key,
                                                  CBS_data(in))) {
+    OPENSSL_free(public_key);
+    OPENSSL_free(private_key);
+    OPENSSL_free(seed);
     OPENSSL_PUT_ERROR(EVP, EVP_R_DECODE_ERROR);
     return 0;
   }
 
+  // copy the seed data
+  OPENSSL_memcpy(seed, CBS_data(in), key->pqdsa->keygen_seed_len);
+
```

Comment:
You shouldn't use memcpy with a CBS, that skips all the nice length safety checks it does internally. We should avoid accessing the underlying buffer using CBS_data unless we actually need the pointer to pass to another function. This should use CBS_copy_bytes.

Reply:
Great call out! Thank you, added in [b482db7](https://github.com/aws/aws-lc/pull/2194/commits/b482db7e346eec1314656871069dfc80fc7c18e7)